/*
 * Telemetry message definitions
 * 
 * Note that all units will be in meters, meters/second, seconds,
 * degrees, etc. unless otherwise noted.
 * 
 * Time is in seconds from 1970 epoch.
 *
 * Yaw is in the range [0, 360) degrees and lat, lon, pitch and roll
 * are in the range (-180, 180].
 */

syntax = "proto3";

package telemetry;

// Represents an overview of the plane's current state.
// Useful for dashboards and flight views.
message Overview {
    double time = 1;
    Position pos = 2;
    Rotation rot = 3;
    Altitude alt = 4;
    Velocity vel = 5;
    Speed speed = 6;
    Battery battery = 7;
}

// WGS84 in degrees
message Position {
    double lat = 1;
    double lon = 2;
}

// In degrees
message Rotation {
    // Semantically eqivalent to heading
    double yaw = 1;
    double pitch = 2;
    double roll = 3;
}

// In meters
message Altitude {
    double msl = 1;
    double agl = 2;
}

// Body frame velocity in m/s
message Velocity {
    double x = 1;
    double y = 2;
    double z = 3;
}

// In m/s
message Speed {
    double airspeed = 1;
    double ground_speed = 2;
}

message Battery {
    // Voltage in Volts.
    double voltage = 1;
    // Current in Amps.
    double current = 2;
    int32 percentage = 3;
}

// Camera telemtry
message CameraTelem {
    double time = 1;
    double lat = 2;
    double lon = 3;
    // Relative to ground
    double alt = 4;
    double yaw = 5;
    // 0 pitch points to the front of the plane, -90 points down
    double pitch = 6;
    // 0 roll points down, 90 points to the right
    double roll = 7;
}

// This has direct mappings to the Mavlink protocol.
message RawMission {
    message RawMissionItem {
        uint32 target_system = 1;
        uint32 target_component = 2;
        // Order of this Command, only needed when reading a
        // RawMission.
        uint32 seq = 3;
        // Frame and Command use the int number specified in the
        // Mavlink XML file.
        uint32 frame = 4;
        uint32 command = 5;
        uint32 current = 6;
        uint32 autocontinue = 7;
        // These parameters correspond to the Mavlink command enums.
        float param_1 = 8;
        float param_2 = 9;
        float param_3 = 10;
        float param_4 = 11;
        float x = 12;
        float y = 13;
        float z = 14;
        uint32 mission_type = 15;
    }

    double time = 1;
    repeated RawMissionItem mission_items = 2;
}

// The current mission item on the plane.
message MissionCurrent {
    double time = 1;
    uint32 item_number = 2;
}
